zebra/netconf_netlink.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * netconf_netlink.c - netconf interaction with the kernel using
+ * netlink
+ * Copyright (C) 2021  Nvidia, Inc.
+ *                     Donald Sharp
+ *
+ * This file is part of FRR.
+ *
+ * FRR is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2, or (at your option) any
+ * later version.
+ *
+ * FRR is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with FRR; see the file COPYING.  If not, write to the Free
+ * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ */
+#include <zebra.h>
+
+#ifdef HAVE_NETLINK /* Netlink OSes only */
+
+#include <ns.h>
+
+#include "linux/netconf.h"
+
+#include "zebra/interface.h"
+#include "zebra/zebra_ns.h"
+#include "zebra/zebra_dplane.h"
+#include "zebra/kernel_netlink.h"
+#include "zebra/netconf_netlink.h"
+#include "zebra/debug.h"
+
+static struct rtattr *netconf_rta(struct netconfmsg *ncm)
+{
+	return (struct rtattr *)((char *)ncm
+				 + NLMSG_ALIGN(sizeof(struct netconfmsg)));
+}
+
+/*
+ * Handle netconf update about a single interface: create dplane
+ * context, and enqueue for processing in the main zebra pthread.
+ */
+static int netlink_netconf_dplane_update(ifindex_t ifindex, bool mpls_on,
+					 bool mcast_on)
+{
+	struct zebra_dplane_ctx *ctx;
+	uint32_t flags = 0;
+
+	ctx = dplane_ctx_alloc();
+	dplane_ctx_set_op(ctx, DPLANE_OP_INTF_NETCONFIG);
+	dplane_ctx_set_netconf_ifindex(ctx, ifindex);
+
+	if (mpls_on)
+		SET_FLAG(flags, DPLANE_NETCONF_MPLS);
+	if (mcast_on)
+		SET_FLAG(flags, DPLANE_NETCONF_MCAST);
+
+	dplane_ctx_set_netconf_flags(ctx, flags);
+
+	/* Enqueue ctx for main pthread to process */
+	dplane_provider_enqueue_to_zebra(ctx);
+
+	return 0;
+}
+
+/*
+ * Parse and process an incoming netlink netconf update.
+ */
+int netlink_netconf_change(struct nlmsghdr *h, ns_id_t ns_id, int startup)
+{
+	struct netconfmsg *ncm;
+	struct rtattr *tb[NETCONFA_MAX + 1] = {};
+	int len;
+	ifindex_t ifindex;
+	uint32_t ival;
+	bool mpls_on = false;
+	bool mc_on = false;
+
+	if (h->nlmsg_type != RTM_NEWNETCONF && h->nlmsg_type != RTM_DELNETCONF)
+		return 0;
+
+	len = h->nlmsg_len - NLMSG_LENGTH(sizeof(struct netconfmsg));
+	if (len < 0) {
+		zlog_err("%s: Message received from netlink is of a broken size: %d, min %zu",
+			 __func__, h->nlmsg_len,
+			 (size_t)NLMSG_LENGTH(sizeof(struct netconfmsg)));
+		return -1;
+	}
+
+	ncm = NLMSG_DATA(h);
+
+	netlink_parse_rtattr(tb, NETCONFA_MAX, netconf_rta(ncm), len);
+
+	if (!tb[NETCONFA_IFINDEX]) {
+		zlog_err("NETCONF message received from netlink without an ifindex");
```

Comment:
`__func__` would be useful, right?

Reply:
not sure - this is the only place we read netlink NETCONF messages, so it's pretty unambiguous?